processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -25,26 +25,29 @@
 import io.druid.data.input.Row;
 import io.druid.query.filter.DimFilter;
 import io.druid.query.filter.ValueMatcher;
-import io.druid.query.groupby.RowBasedValueMatcherFactory;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
 
 import java.nio.ByteBuffer;
+import java.util.Map;
 
-public class DimFilterHavingSpec implements HavingSpec
+public class DimFilterHavingSpec extends BaseHavingSpec
 {
   private static final byte CACHE_KEY = (byte) 0x9;
 
   private final DimFilter dimFilter;
-  private final RowBasedValueMatcherFactory valueMatcherFactory;
-  private final ValueMatcher valueMatcher;
+  private final ThreadLocal<Row> rowSupplier;
+
+  private final Object valueMatcherMonitor = new Object();
+  private volatile ValueMatcher valueMatcher;
```

Comment:
@leventov this is to better understand https://github.com/druid-io/druid/pull/3797#discussion_r93802927 , say, we have
```
private volatile ValueMatcher valueMatcher;
```

now, at time t1, Thread A does a write on that...

```
valueMather = new ValueMatcher() { ...} 

```


and, at time t2 greater than t1, Thread B reads same...

```
if (valueMatcher == null) {
   System.out.println("I am null");
}
```

are you saying it is possible that "I am null" might get printed at t2 and afterwards ?


that sounds against the "volatile variable rule".... "A write to a volatile field happens-before every subsequent read of that field."